From: Matthew Hagemann <matt.hagemann@canonical.com>
Date: Thu, 13 Feb 2025 17:08:34 +0200
Subject: Dynamic lang pack installation support

Adding a polkit rule to allow the gnome initial setup user to run pkexec if the
program it is running is apt. This is to allow for Gnome-Initial-Setup install
language packs during setup via apt update && apt install.

Adding a function apt_install that takes in a list of packages and attempts to
install them. This is inteded to be used as part of the new flow for
dynamically installing language packs for the user during setup.

Added parse_language_list function to parse the static list of languages in
/usr/share/language-selector/data/languagelist which is now used to populate
the language chooser.

Added the ability for GIS to determine what outstanding language packs need to
be installed via check-language-support -l and to use that output to install
outstanding lang packs via a g_task.

Added a spinner to the language chooser for when a language's packs are being
installed. Language chooser and next button are both set to sensitive FALSE
while GIS attempts to install language packs. If language pack installation
fails, a warning is shown to the user and the locale is set.
---
 data/20-gnome-initial-setup.rules.in               |  10 +
 gnome-initial-setup/gis-pkexec.c                   |  47 +++++
 gnome-initial-setup/gis-pkexec.h                   |   4 +
 .../pages/language/cc-language-chooser.c           | 230 +++++++++++++++++++--
 .../pages/language/gis-language-page.c             |  22 +-
 5 files changed, 292 insertions(+), 21 deletions(-)

diff --git a/data/20-gnome-initial-setup.rules.in b/data/20-gnome-initial-setup.rules.in
index 02fd21d..f0faf42 100644
--- a/data/20-gnome-initial-setup.rules.in
+++ b/data/20-gnome-initial-setup.rules.in
@@ -5,6 +5,16 @@
 // Allow the gnome-initial-setup user to do certain actions without
 // being interrupted by password dialogs
 
+polkit.addRule(function(action, subject) {
+    if (subject.user !== 'gnome-initial-setup')
+        return undefined;
+
+    if (action.id === 'org.freedesktop.policykit.exec' &&
+        action.lookup('program') === '/usr/bin/apt') {
+        return 'yes';
+    }
+});
+
 polkit.addRule(function(action, subject) {
     if (subject.user !== 'gnome-initial-setup')
         return undefined;
diff --git a/gnome-initial-setup/gis-pkexec.c b/gnome-initial-setup/gis-pkexec.c
index 9e1bfa9..3409682 100644
--- a/gnome-initial-setup/gis-pkexec.c
+++ b/gnome-initial-setup/gis-pkexec.c
@@ -24,6 +24,7 @@
  */
 
 #include "gis-pkexec.h"
+#include "glib.h"
 
 gboolean
 gis_pkexec (const gchar  *command,
@@ -57,3 +58,49 @@ gis_pkexec (const gchar  *command,
 
   return TRUE;
 }
+
+gboolean apt_install(GPtrArray *packages, GError **error) {
+  g_autoptr(GSubprocessLauncher) launcher = NULL;
+  g_autoptr(GSubprocess) process = NULL;
+
+  if (!packages || packages->len == 0) {
+      g_warning("List of packages to install is NULL or empty.");
+      return FALSE;
+  }
+
+  /* Create space for 8 leading arguments of the install command, the number of 
+   * packages and a NULL terminator. 
+   */
+  int total = 4 + packages->len + 1;
+  gchar **argv = g_new(gchar *, total);
+  int pos = 0;
+  argv[pos++] = "/usr/bin/pkexec";
+  argv[pos++] = "apt";
+  argv[pos++] = "install";
+  argv[pos++] = "-y";
+  for (int i = 0; i < packages->len; i++)
+    argv[pos++] = (gchar *)g_ptr_array_index(packages, i);
+  argv[pos] = NULL;
+
+  launcher = g_subprocess_launcher_new(G_SUBPROCESS_FLAGS_NONE);
+  g_subprocess_launcher_unsetenv(launcher, "SHELL");
+
+  g_subprocess_launcher_setenv(launcher, "DEBCONF_NONINTERACTIVE_SEEN", "true", TRUE);
+  g_subprocess_launcher_setenv(launcher, "DEBIAN_FRONTEND", "noninteractive", TRUE);
+
+  process =
+      g_subprocess_launcher_spawnv(launcher, (const gchar *const *)argv, error);
+  g_free(argv);
+
+  if (!process) {
+    g_prefix_error(error, "Failed to create apt install process: ");
+    return FALSE;
+  }
+
+  if (!g_subprocess_wait_check(process, NULL, error)) {
+    g_prefix_error(error, "apt install failed: ");
+    return FALSE;
+  }
+
+  return TRUE;
+}
diff --git a/gnome-initial-setup/gis-pkexec.h b/gnome-initial-setup/gis-pkexec.h
index ff5a88c..74cdeb2 100644
--- a/gnome-initial-setup/gis-pkexec.h
+++ b/gnome-initial-setup/gis-pkexec.h
@@ -36,6 +36,10 @@ gis_pkexec (const gchar  *command,
             const gchar  *user,
             GError      **error);
 
+gboolean
+apt_install (GPtrArray *packages, 
+                 GError **error);
+
 G_END_DECLS
 
 #endif /* __GIS_PKEXEC_H__ */
diff --git a/gnome-initial-setup/pages/language/cc-language-chooser.c b/gnome-initial-setup/pages/language/cc-language-chooser.c
index 61756be..c7e0553 100644
--- a/gnome-initial-setup/pages/language/cc-language-chooser.c
+++ b/gnome-initial-setup/pages/language/cc-language-chooser.c
@@ -21,6 +21,8 @@
  */
 
 #include "config.h"
+#include "gis-pkexec.h"
+#include "glib.h"
 #include "cc-language-chooser.h"
 
 #include <locale.h>
@@ -41,7 +43,7 @@ struct _CcLanguageChooserPrivate
 {
         GtkWidget *filter_entry;
         GtkWidget *language_list;
-
+        
         GtkWidget *no_results;
         GtkWidget *more_item;
 
@@ -69,8 +71,10 @@ static guint signals[LAST_SIGNAL] = { 0 };
 
 typedef struct {
         GtkWidget *box;
+        GtkWidget *status_stack;
         GtkWidget *checkmark;
-
+        GtkWidget *spinner;
+        GtkWidget *status_label;
         gchar *locale_id;
         gchar *locale_name;
         gchar *locale_current_name;
@@ -155,8 +159,18 @@ language_widget_new (const char *locale_id,
         gtk_label_set_xalign (GTK_LABEL (label), 0);
         gtk_box_append (GTK_BOX (widget->box), label);
 
+        widget->status_stack = gtk_stack_new();
         widget->checkmark = gtk_image_new_from_icon_name ("object-select-symbolic");
-        gtk_box_append (GTK_BOX (widget->box), widget->checkmark);
+        widget->spinner = gtk_spinner_new();
+        gtk_stack_add_titled(GTK_STACK(widget->status_stack), widget->checkmark, "check", "Checkmark");
+        gtk_stack_add_titled(GTK_STACK(widget->status_stack), widget->spinner, "spinner", "Spinner");
+        widget->status_label = gtk_label_new("");
+        gtk_widget_set_name(widget->status_label, "status");
+        gtk_widget_add_css_class(widget->status_label, "dim-label");
+
+        gtk_stack_set_visible_child_name(GTK_STACK(widget->status_stack), "check");
+        gtk_box_append(GTK_BOX(widget->box), widget->status_stack);
+        gtk_box_append(GTK_BOX(widget->box), widget->status_label);
 
         if (country_name) {
                 label = gtk_label_new (country_name);
@@ -188,7 +202,9 @@ language_widget_new (const char *locale_id,
 }
 
 static void
-sync_all_checkmarks (CcLanguageChooser *chooser)
+sync_all_checkmarks (CcLanguageChooser *chooser, 
+                     gboolean waiting, 
+                     const gchar *error_message)
 {
         CcLanguageChooserPrivate *priv = cc_language_chooser_get_instance_private (chooser);
         GtkWidget *row;
@@ -206,10 +222,28 @@ sync_all_checkmarks (CcLanguageChooser *chooser)
                         return;
 
                 should_be_visible = g_str_equal (widget->locale_id, priv->language);
+                if (should_be_visible) {
+                        if (waiting) {
+                                gtk_stack_set_visible_child_name(GTK_STACK(widget->status_stack), "spinner");
+                                gtk_spinner_start(GTK_SPINNER(widget->spinner));
+                                gtk_label_set_text(GTK_LABEL(widget->status_label), _("Downloading language packs..."));
+                    } else {
+                            gtk_spinner_stop(GTK_SPINNER(widget->spinner));
+                            gtk_stack_set_visible_child_name(GTK_STACK(widget->status_stack), "check");
+                            if (error_message) {
+                                    gtk_label_set_text(GTK_LABEL(widget->status_label), error_message);
+                            } else {
+                                    gtk_label_set_text(GTK_LABEL(widget->status_label), "");
+                            }
+                    }
+                } else {
+                        gtk_stack_set_visible_child_name(GTK_STACK(widget->status_stack), "check");
+                        gtk_label_set_text(GTK_LABEL(widget->status_label), "");
+                } 
+                        
                 gtk_widget_set_opacity (widget->checkmark, should_be_visible ? 1.0 : 0.0);
-
                 row = gtk_widget_get_next_sibling (row);
-        }
+            }
 }
 
 static GtkWidget *
@@ -290,13 +324,60 @@ add_languages (CcLanguageChooser  *chooser,
         gtk_list_box_set_placeholder (GTK_LIST_BOX (priv->language_list), priv->no_results);
 }
 
+gchar **
+parse_language_list(void)
+{
+    const gchar *filepath = "/usr/share/language-selector/data/languagelist";
+    gchar *contents = NULL;
+    gsize length;
+    GError *error = NULL;
+
+    if (!g_file_get_contents(filepath, &contents, &length, &error)) {
+        g_printerr("Error reading file: %s\n", error->message);
+        g_error_free(error);
+        return NULL;
+    }
+
+    gchar **lines = g_strsplit(contents, "\n", -1);
+    g_free(contents);
+
+    GPtrArray *codes = g_ptr_array_new_with_free_func(g_free);
+    for (gchar **line = lines; *line != NULL; line++) {
+        gchar *trimmed = g_strstrip(*line);
+        
+        /* Skip empty lines and commented out lines */
+        if (trimmed[0] == '\0' || trimmed[0] == '#') {
+            continue;
+        }
+
+        gchar **tokens = g_strsplit(trimmed, ";", -1);
+        gint token_count = g_strv_length(tokens);
+
+        /* Language code we use is the 4th entry in a ; delimited list */
+        if (token_count >= 4) {
+            g_ptr_array_add(codes, g_strdup(tokens[3]));
+        }
+        g_strfreev(tokens);
+    }
+    g_strfreev(lines);
+
+    g_ptr_array_add(codes, NULL);
+
+    /* gchar ** is the required type of the function consuming this output */
+    return (gchar **) g_ptr_array_free(codes, FALSE);
+}
+
 static void
 add_all_languages (CcLanguageChooser *chooser)
 {
         g_auto(GStrv) locale_ids = NULL;
+
+        /* Make use of the static language list in 
+         * /usr/share/language-selector/data/languagelist
+         */
+        locale_ids = parse_language_list();
         g_autoptr(GHashTable) initial = NULL;
 
-        locale_ids = gnome_get_all_locales ();
         initial = cc_common_language_get_initial_languages ();
         add_languages (chooser, locale_ids, initial);
 }
@@ -395,21 +476,129 @@ show_more (CcLanguageChooser *chooser)
         g_object_notify_by_pspec (G_OBJECT (chooser), obj_props[PROP_SHOWING_EXTRA]);
 }
 
+static gchar *
+get_lang_support_output(const gchar *locale)
+{
+  g_autoptr(GSubprocess) subprocess = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autofree gchar *stdout_buf = NULL;
+  g_autofree gchar *stderr_buf = NULL;
+
+  subprocess = g_subprocess_new(
+    G_SUBPROCESS_FLAGS_STDOUT_PIPE | G_SUBPROCESS_FLAGS_STDERR_PIPE,
+    &error,
+    "/usr/bin/check-language-support",
+    "-l",
+    locale,
+    NULL
+  );
+
+  if (!subprocess) {
+      g_warning("Failed to create subprocess: %s", error->message);
+      return NULL;
+  }
+
+  if (!g_subprocess_communicate_utf8(subprocess, NULL, NULL, &stdout_buf, &stderr_buf, &error)) {
+      g_warning("Failed to communicate with subprocess: %s", error->message);
+      return NULL;
+  }
+
+  if (stderr_buf && *stderr_buf != '\0') {
+    g_warning("check-language-support stderr: %s", stderr_buf);
+  }
+
+  return g_strdup(stdout_buf);
+}
+
+gboolean
+install_language_support(const gchar *locale) {
+  g_autofree gchar *packages_str = get_lang_support_output(locale);
+  GError *error = NULL;
+
+  if (!packages_str || g_strcmp0(packages_str, "") == 0) {
+     g_warning("No packages to install for locale: %s", locale);
+     return TRUE;
+  }
+
+  gchar** packages_split = g_strsplit(packages_str, " ", -1);
+  GPtrArray *packages_array = g_ptr_array_new_with_free_func(g_free);
+
+  for (int i = 0; packages_split[i] != NULL; i++) {
+     gchar *trimmed = g_strstrip(packages_split[i]);
+     if (*trimmed != '\0')
+       g_ptr_array_add(packages_array, g_strdup(trimmed));
+  }
+  g_strfreev(packages_split);
+
+  if (packages_array->len == 0) {
+     g_warning("No valid packages to install for locale: %s", locale);
+     g_ptr_array_free(packages_array, TRUE);
+     return TRUE;
+  }
+
+  if (!apt_install(packages_array, &error)) {
+     g_printerr("Error installing language support: %s\n", error->message);
+     g_clear_error(&error);
+     return FALSE;
+  }
+
+  g_ptr_array_free(packages_array, TRUE);
+  return TRUE;
+}
+
+static void
+installation_task_async (GTask      *task,
+                         gpointer    source_object,
+                         gpointer    task_data,
+                         GCancellable *cancellable)
+{
+  const gchar *locale = task_data;
+  gboolean success = install_language_support(locale);
+  g_task_return_boolean(task, success);
+}
+
+static void
+installation_task_done (GObject    *source_object,
+                        GAsyncResult *res,
+                        gpointer    user_data)
+{
+  CcLanguageChooser *chooser = user_data;
+  gboolean success = g_task_propagate_boolean(G_TASK(res), NULL);
+  if (!success) {
+    sync_all_checkmarks(chooser, false, _("Unable to download additional language packs"));
+  } else {
+    sync_all_checkmarks(chooser, false, NULL);
+  } 
+
+  gtk_widget_set_sensitive(GTK_WIDGET(chooser), TRUE);
+}
+
 static void
 set_locale_id (CcLanguageChooser *chooser,
                const gchar       *new_locale_id)
 {
-        CcLanguageChooserPrivate *priv = cc_language_chooser_get_instance_private (chooser);
+  CcLanguageChooserPrivate *priv = cc_language_chooser_get_instance_private(chooser);
 
-        if (g_strcmp0 (priv->language, new_locale_id) == 0)
-                return;
+  if (g_strcmp0(priv->language, new_locale_id) == 0)
+      return;
 
-        g_free (priv->language);
-        priv->language = g_strdup (new_locale_id);
+  g_free(priv->language);
+  priv->language = g_strdup(new_locale_id);
+
+  sync_all_checkmarks(chooser, true, NULL);
+    
+  gchar *locale_for_install = NULL;
+  if (g_str_has_suffix(new_locale_id, ".UTF-8"))
+    locale_for_install = g_strndup(new_locale_id, strlen(new_locale_id) - strlen(".UTF-8"));
+  else
+    locale_for_install = g_strdup(new_locale_id);
 
-        sync_all_checkmarks (chooser);
+  GTask *task = g_task_new(chooser, NULL, installation_task_done, chooser);
+  g_task_set_task_data(task, locale_for_install, g_free);
+  g_task_run_in_thread(task, installation_task_async);
+  g_object_unref(task);
 
-        g_object_notify_by_pspec (G_OBJECT (chooser), obj_props[PROP_LANGUAGE]);
+  g_object_notify_by_pspec(G_OBJECT(chooser), obj_props[PROP_LANGUAGE]);
 }
 
 static gboolean
@@ -441,10 +630,13 @@ row_activated (GtkListBox        *box,
                 widget = get_language_widget (child);
                 if (widget == NULL)
                         return;
-                if (g_strcmp0 (priv->language, widget->locale_id) == 0)
-			g_idle_add (confirm_choice, chooser);
-                else
-                        set_locale_id (chooser, widget->locale_id);
+                if (g_strcmp0 (priv->language, widget->locale_id) == 0) {
+			        g_idle_add (confirm_choice, chooser);
+                }
+                else {
+                    gtk_widget_set_sensitive(GTK_WIDGET(chooser), FALSE);
+                    set_locale_id (chooser, widget->locale_id);
+                }
         }
 }
 
@@ -477,7 +669,7 @@ cc_language_chooser_constructed (GObject *object)
         if (priv->language == NULL)
                 priv->language = cc_common_language_get_current_language ();
 
-        sync_all_checkmarks (chooser);
+        sync_all_checkmarks (chooser, false, NULL);
 }
 
 static void
diff --git a/gnome-initial-setup/pages/language/gis-language-page.c b/gnome-initial-setup/pages/language/gis-language-page.c
index fe5fed9..98013ae 100644
--- a/gnome-initial-setup/pages/language/gis-language-page.c
+++ b/gnome-initial-setup/pages/language/gis-language-page.c
@@ -24,6 +24,8 @@
 
 /* Language page {{{1 */
 
+#include "glib.h"
+#include <unistd.h>
 #define PAGE_ID "language"
 
 #define GNOME_SYSTEM_LOCALE_DIR "org.gnome.system.locale"
@@ -56,6 +58,17 @@ typedef struct _GisLanguagePagePrivate GisLanguagePagePrivate;
 
 G_DEFINE_TYPE_WITH_PRIVATE (GisLanguagePage, gis_language_page, GIS_TYPE_PAGE);
 
+static void
+language_chooser_sensitive_changed (GObject      *object,
+                                    GParamSpec   *pspec,
+                                    gpointer      user_data)
+{
+  GisLanguagePage *page = GIS_LANGUAGE_PAGE (user_data);
+  GisLanguagePagePrivate *priv = gis_language_page_get_instance_private (page);
+  gboolean complete = gtk_widget_get_sensitive (priv->language_chooser);
+  gis_page_set_complete (GIS_PAGE (page), complete);
+}
+
 static void
 set_localed_locale (GisLanguagePage *self)
 {
@@ -250,8 +263,13 @@ gis_language_page_constructed (GObject *object)
                         object);
       g_object_unref (bus);
     }
-
-  gis_page_set_complete (GIS_PAGE (page), TRUE);
+  
+  /* Block the user from advancing until the language chooser has completed
+  *  updating the locale.
+  */
+  gis_page_set_complete (GIS_PAGE (page), gtk_widget_get_sensitive (priv->language_chooser));
+  g_signal_connect (priv->language_chooser, "notify::sensitive",
+                    G_CALLBACK (language_chooser_sensitive_changed), page);
   gtk_widget_set_visible (GTK_WIDGET (page), TRUE);
 }
 
